### PR TITLE
[CQT-237] Create CNOT to CZ replacer

### PR DESCRIPTION
**Pull request by elenbaasc**
- Implement CNOT2CZ decomposer
- Add tests for CNOT2CZ decomposer
- Adjust output of backend integration tests

### CODE REVIEW

File: opensquirrel/passes/decomposer/cnot2cz_decomposer.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+class CNOT2CZDecomposer(Decomposer):
+    """Predefined decomposition of CNOT gate to CZ gate with Y rotations.
+
+    ---•---     -----------------•----------------
+       |     →                   |
+    ---⊕---     --[Ry(-pi/2)]---[Z]---[Ry(pi/2)]--
+
+    Note:
+        This decomposition preserves the global phase of the CNOT gate.
+
```
Comment (rturrado) [inline opensquirrel/passes/decomposer/cnot2cz_decomposer.py:19]:
```suggestion
```

File: opensquirrel/passes/decomposer/cnot2cz_decomposer.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            return [gate]
+        if isinstance(gate, ControlledGate):
+            if not isinstance(gate.target_gate, BlochSphereRotation):
+                # Do nothing.
+                # ControlledGate's with 2+ control qubits are ignored.
+                return [gate]
+            control_qubit = gate.control_qubit
+            target_qubit = gate.target_gate.qubit
+        else:
+            # If CNOT is not implemented as a ControlledGate, but e.g. as a MatrixGate.
```
Comment (rturrado) [inline opensquirrel/passes/decomposer/cnot2cz_decomposer.py:34]:
```suggestion
            # If CNOT is not implemented as a ControlledGate but, e.g., as a MatrixGate.
```

File: opensquirrel/passes/decomposer/cnot2cz_decomposer.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    ---⊕---     --[Ry(-pi/2)]---[Z]---[Ry(pi/2)]--
+
+    Note:
+        This decomposition preserves the global phase of the CNOT gate.
+
+    """
+
+    def decompose(self, gate: Gate) -> list[Gate]:
+        if gate.name != "CNOT":
+            # Do nothing.
```
Comment (rturrado) [inline opensquirrel/passes/decomposer/cnot2cz_decomposer.py:24]: I wouldn't add this comment.

```suggestion
```

File: opensquirrel/passes/decomposer/cnot2cz_decomposer.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            return [gate]
+        if isinstance(gate, ControlledGate):
+            if not isinstance(gate.target_gate, BlochSphereRotation):
+                # Do nothing.
```
Comment (rturrado) [inline opensquirrel/passes/decomposer/cnot2cz_decomposer.py:28]: Same.

```suggestion
```

File: test/test_integration.py [hunk 5]
```diff
@@ -6,13 +6,12 @@
...
 from opensquirrel.circuit import Circuit
-from opensquirrel.default_instructions import CNOT, CZ, H
 from opensquirrel.ir import Measure
-from opensquirrel.passes.decomposer import CNOTDecomposer, McKayDecomposer, XYXDecomposer
+from opensquirrel.passes.decomposer import CNOT2CZDecomposer, CNOTDecomposer, McKayDecomposer, XYXDecomposer
 from opensquirrel.passes.exporter.export_format import ExportFormat
 
 
-def test_Spin2_backend() -> None:  # noqa: N802
+def test_spin_backend() -> None:
```
Comment (rturrado) [inline test/test_integration.py:14]: Do we want to go from `Spin2` to `spin`?

```suggestion
def test_spin_2_backend() -> None:
```

**Review by rturrado — COMMENTED** [review]
Really nice!

Do we want to update any docs?

File: test/test_integration.py (hunk 5 above)
Comment (elenbaasc) [reply test/test_integration.py:14]: I left out the 2 on purpose; this integration test is valid for the spin system in general (the number of qubits is irrelevant). So also valid for the Spin-6 and later the Spin-8 system.

**Comment by elenbaasc** [comment]
> Really nice!
> 
> Do we want to update any docs?

Yeah, good point. I'll have a look at this.

File: test/test_integration.py (hunk 5 above)
Comment (rturrado) [reply test/test_integration.py:14]: Great!

File: docs/tutorial.md [hunk 6]
```diff
@@ -368,20 +368,30 @@ _Output_:
...
+##### _CNOT to CZ decomposer_
+
+The decomposition of the `CNOT` gate into a `CZ` gate (with additional single-qubit gates) is used frequently.
+To this end a `CNOT2CZDecomposer` has been implemented that decomposes any `CNOT`s in a circuit to a
+`Ry(-π/2)`-`CZ`-`Ry(π/2)`; the decomposition is illustrated in the image below.
+
+<p align="center"> <img width="600" src="_static/cnot2cz.png"> </p>
+
+`Ry` gates are used instead of, _e.g._, `H` gates,
+as they are generally already more likely to be supported by target backends.
```
Comment (rturrado) [inline docs/tutorial.md]: Suggestion: `as they are, generally, more likely to be supported already by target backends.`

`Generally already more likely` sounds a bit weird to me.

File: docs/tutorial.md [hunk 7]
```diff
@@ -368,20 +368,30 @@ _Output_:
 
     replacement for gate CNOT does not preserve the quantum state
 
+##### _CNOT to CZ decomposer_
+
+The decomposition of the `CNOT` gate into a `CZ` gate (with additional single-qubit gates) is used frequently.
+To this end a `CNOT2CZDecomposer` has been implemented that decomposes any `CNOT`s in a circuit to a
+`Ry(-π/2)`-`CZ`-`Ry(π/2)`; the decomposition is illustrated in the image below.
```
Comment (rturrado) [inline docs/tutorial.md:376]: Suggestion: `. The decomposition`.

Very personal. I don't really like semicolons much haha

File: docs/tutorial.md (hunk 6 above)
Comment (rturrado) [reply docs/tutorial.md]: Ah, I see  you changed this later. Maybe you still want to add the `already`.

**Review by rturrado — DISMISSED** [review]

**Review by rturrado — APPROVED** [review]
Nice work!

**Pull request merged**